File: backend/store_test.go
package main

import (
	"context"
	"testing"
	"time"

	"github.com/redis/go-redis/v9"
)

func TestRedisStorePutGet(t *testing.T) {
	store := testRedisStore(t)
	ctx := context.Background()

	meta := SessionMeta{
		ID:           "sess-1",
		UserID:       "alice@example.com",
		Directory:    "/home/alice/project",
		TunnelerAddr: "10.0.0.1:9090",
		CreatedAt:    time.Now().Truncate(time.Millisecond),
	}

	if err := store.Put(ctx, meta); err != nil {
		t.Fatal(err)
	}

	got, ok, err := store.Get(ctx, "sess-1")
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected session to exist")
	}
	if got.UserID != "alice@example.com" {
		t.Errorf("UserID = %q, want %q", got.UserID, "alice@example.com")
	}
	if got.TunnelerAddr != "10.0.0.1:9090" {
		t.Errorf("TunnelerAddr = %q, want %q", got.TunnelerAddr, "10.0.0.1:9090")
	}
}

func TestRedisStoreGetMissing(t *testing.T) {
	store := testRedisStore(t)
	_, ok, err := store.Get(context.Background(), "nonexistent")
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected session to not exist")
	}
}

func TestRedisStoreDelete(t *testing.T) {
	store := testRedisStore(t)
	ctx := context.Background()

	store.Put(ctx, SessionMeta{ID: "sess-1", UserID: "alice@example.com"})
	if err := store.Delete(ctx, "sess-1"); err != nil {
		t.Fatal(err)
	}

	_, ok, _ := store.Get(ctx, "sess-1")
	if ok {
		t.Fatal("expected session to be deleted")
	}
}

func TestRedisStoreList(t *testing.T) {
	store := testRedisStore(t)
	ctx := context.Background()

	store.Put(ctx, SessionMeta{ID: "sess-1", UserID: "alice@example.com"})
	store.Put(ctx, SessionMeta{ID: "sess-2", UserID: "alice@example.com"})
	store.Put(ctx, SessionMeta{ID: "sess-3", UserID: "bob@example.com"})

	sessions, err := store.List(ctx, "alice@example.com")
	if err != nil {
		t.Fatal(err)
	}
	if len(sessions) != 2 {
		t.Errorf("got %d sessions, want 2", len(sessions))
	}
}

func TestRedisStorePutOverwrite(t *testing.T) {
	store := testRedisStore(t)
	ctx := context.Background()

	store.Put(ctx, SessionMeta{ID: "sess-1", UserID: "alice@example.com", Directory: "/old"})
	store.Put(ctx, SessionMeta{ID: "sess-1", UserID: "alice@example.com", Directory: "/new"})

	got, _, _ := store.Get(ctx, "sess-1")
	if got.Directory != "/new" {
		t.Errorf("Directory = %q, want %q", got.Directory, "/new")
	}
}

func TestRedisStoreTTL(t *testing.T) {
	client := testRedisClient(t)
	store := NewRedisStore(client, 1*time.Second)
	ctx := context.Background()

	store.Put(ctx, SessionMeta{ID: "sess-ttl", UserID: "alice@example.com"})
	_, ok, _ := store.Get(ctx, "sess-ttl")
	if !ok {
		t.Fatal("expected session to exist before TTL")
	}

	time.Sleep(1500 * time.Millisecond)

	_, ok, _ = store.Get(ctx, "sess-ttl")
	if ok {
		t.Fatal("expected session to be expired after TTL")
	}
}

func TestRedisStorePutRefreshesTTL(t *testing.T) {
	client := testRedisClient(t)
	store := NewRedisStore(client, 2*time.Second)
	ctx := context.Background()

	store.Put(ctx, SessionMeta{ID: "sess-refresh", UserID: "alice@example.com"})
	time.Sleep(1 * time.Second)

	store.Put(ctx, SessionMeta{ID: "sess-refresh", UserID: "alice@example.com"})
	time.Sleep(1500 * time.Millisecond)

	_, ok, _ := store.Get(ctx, "sess-refresh")
	if !ok {
		t.Fatal("expected session to still exist after TTL refresh")
	}
}

func TestRedisStorePing(t *testing.T) {
	store := testRedisStore(t)
	if err := store.Ping(context.Background()); err != nil {
		t.Fatal(err)
	}
}

func TestRedisStoreGetCorruptData(t *testing.T) {
	client := testRedisClient(t)
	store := NewRedisStore(client, 10*time.Minute)
	ctx := context.Background()

	// Put corrupt data directly into Redis
	client.Set(ctx, "orc:session:corrupt-sess", "not-valid-json{{{", 10*time.Minute)

	_, _, err := store.Get(ctx, "corrupt-sess")
	if err == nil {
		t.Fatal("expected error for corrupt session data")
	}
}

func TestRedisStoreListWithCorruptEntry(t *testing.T) {
	client := testRedisClient(t)
	store := NewRedisStore(client, 10*time.Minute)
	ctx := context.Background()

	// Add a valid session
	store.Put(ctx, SessionMeta{ID: "good-sess", UserID: "alice@example.com"})

	// Corrupt one entry directly
	client.Set(ctx, "orc:session:bad-sess", "not-json{{{", 10*time.Minute)
	client.SAdd(ctx, "orc:user-sessions:alice@example.com", "bad-sess")

	sessions, err := store.List(ctx, "alice@example.com")
	if err != nil {
		t.Fatal(err)
	}
	// Should only return the good session (corrupt one is skipped)
	if len(sessions) != 1 {
		t.Errorf("expected 1 valid session, got %d", len(sessions))
	}
}

func TestRedisStoreBrokenConnection(t *testing.T) {
	client := redis.NewClient(&redis.Options{Addr: "localhost:1"})
	store := NewRedisStore(client, 10*time.Minute)
	ctx := context.Background()

	if err := store.Put(ctx, SessionMeta{ID: "s1", UserID: "u1"}); err == nil {
		t.Error("expected error from Put on broken connection")
	}
	if _, _, err := store.Get(ctx, "s1"); err == nil {
		t.Error("expected error from Get on broken connection")
	}
	if err := store.Delete(ctx, "s1"); err == nil {
		t.Error("expected error from Delete on broken connection")
	}
	if _, err := store.List(ctx, "u1"); err == nil {
		t.Error("expected error from List on broken connection")
	}
}

func TestRedisStoreListCleansStaleIDs(t *testing.T) {
	client := testRedisClient(t)
	store := NewRedisStore(client, 10*time.Minute)
	ctx := context.Background()

	store.Put(ctx, SessionMeta{ID: "sess-1", UserID: "alice@example.com"})
	store.Put(ctx, SessionMeta{ID: "sess-2", UserID: "alice@example.com"})

	client.Del(ctx, "orc:session:sess-2")

	sessions, err := store.List(ctx, "alice@example.com")
	if err != nil {
		t.Fatal(err)
	}
	if len(sessions) != 1 {
		t.Errorf("got %d sessions, want 1 (stale cleaned)", len(sessions))
	}

	members, _ := client.SMembers(ctx, "orc:user-sessions:alice@example.com").Result()
	if len(members) != 1 {
		t.Errorf("stale ID not cleaned from set: got %d members", len(members))
	}
}

File: api/src/store.test.ts
import { describe, it, expect, beforeEach } from "vitest";
import { SessionStore, type SessionMeta } from "./store.js";

class MockRedis {
  private data = new Map<string, string>();
  private sets = new Map<string, Set<string>>();

  async get(key: string): Promise<string | null> {
    return this.data.get(key) ?? null;
  }

  async mget(...keys: string[]): Promise<(string | null)[]> {
    return keys.map((k) => this.data.get(k) ?? null);
  }

  async smembers(key: string): Promise<string[]> {
    return Array.from(this.sets.get(key) ?? []);
  }

  async srem(key: string, ...members: string[]): Promise<number> {
    const set = this.sets.get(key);
    if (!set) return 0;
    let count = 0;
    for (const m of members) {
      if (set.delete(m)) count++;
    }
    return count;
  }

  async ping(): Promise<string> {
    return "PONG";
  }

  putSession(meta: SessionMeta) {
    this.data.set(`orc:session:${meta.id}`, JSON.stringify(meta));
    const setKey = `orc:user-sessions:${meta.userId}`;
    if (!this.sets.has(setKey)) this.sets.set(setKey, new Set());
    this.sets.get(setKey)!.add(meta.id);
  }

  addStaleRef(userId: string, sessionId: string) {
    const setKey = `orc:user-sessions:${userId}`;
    if (!this.sets.has(setKey)) this.sets.set(setKey, new Set());
    this.sets.get(setKey)!.add(sessionId);
  }
}

describe("SessionStore", () => {
  let redis: MockRedis;
  let store: SessionStore;

  beforeEach(() => {
    redis = new MockRedis();
    store = new SessionStore(redis as any);
  });

  describe("get", () => {
    it("returns null for missing session", async () => {
      expect(await store.get("nonexistent")).toBeNull();
    });

    it("returns session when present", async () => {
      redis.putSession({
        id: "sess-1",
        userId: "alice@example.com",
        directory: "/home/alice/project",
        tunnelerAddr: "tunneler:9090",
        createdAt: "2026-01-01T00:00:00Z",
      });
      const result = await store.get("sess-1");
      expect(result).not.toBeNull();
      expect(result!.id).toBe("sess-1");
      expect(result!.userId).toBe("alice@example.com");
      expect(result!.directory).toBe("/home/alice/project");
    });
  });

  describe("list", () => {
    it("returns empty array when user has no sessions", async () => {
      const result = await store.list("nobody@example.com");
      expect(result).toEqual([]);
    });

    it("returns all sessions for user", async () => {
      redis.putSession({
        id: "sess-1",
        userId: "alice@example.com",
        directory: "/project-a",
        tunnelerAddr: "t:9090",
        createdAt: "2026-01-01T00:00:00Z",
      });
      redis.putSession({
        id: "sess-2",
        userId: "alice@example.com",
        directory: "/project-b",
        tunnelerAddr: "t:9090",
        createdAt: "2026-01-01T00:00:00Z",
      });
      const result = await store.list("alice@example.com");
      expect(result).toHaveLength(2);
      const ids = result.map((s) => s.id).sort();
      expect(ids).toEqual(["sess-1", "sess-2"]);
    });

    it("cleans up stale references", async () => {
      redis.putSession({
        id: "sess-1",
        userId: "alice@example.com",
        directory: "/project",
        tunnelerAddr: "t:9090",
        createdAt: "2026-01-01T00:00:00Z",
      });
      redis.addStaleRef("alice@example.com", "sess-gone");

      const result = await store.list("alice@example.com");
      expect(result).toHaveLength(1);
      expect(result[0].id).toBe("sess-1");

      const members = await redis.smembers("orc:user-sessions:alice@example.com");
      expect(members).not.toContain("sess-gone");
    });

    it("does not return other users sessions", async () => {
      redis.putSession({
        id: "sess-1",
        userId: "alice@example.com",
        directory: "/project",
        tunnelerAddr: "t:9090",
        createdAt: "2026-01-01T00:00:00Z",
      });
      const result = await store.list("bob@example.com");
      expect(result).toEqual([]);
    });
  });

  describe("ping", () => {
    it("returns true when redis responds", async () => {
      expect(await store.ping()).toBe(true);
    });

    it("returns false when redis throws", async () => {
      const broken = {
        ping: async () => { throw new Error("connection refused"); },
      };
      const brokenStore = new SessionStore(broken as any);
      expect(await brokenStore.ping()).toBe(false);
    });
  });
});
